src/mam4xx/modal_aer_opt.hpp

```diff
@@ -0,0 +1,1084 @@
+  /*------------------------------------------------------------------------------
+   Bilinear interpolation along the refractive index dimensions
+   of the table to estimate Chebyshev coefficients at an
+   intermediate refractive index.
+
+   In short wave, the first call computes itab, jtab, ttab, utab and coef.
+   The subsequent calls use itab, jtab, ttab and utab as inputs and compute coef
+
+   In long wave, we have just one call to compute itab,jtab,ttab, utab and coef
+  ------------------------------------------------------------------------------*/
+
+  // intent-ins
+  // ncol
+  // table(ncoef,prefr,prefi)
+  // ref_real(pcols), ref_img(pcols) !real and imganinary parts of refractive
+  // indices [unitless] ref_real_tab(prefr), ref_img_tab(prefi) !real and
+  // imganinary table refractive indices [unitless]
+
+  // intent-inouts/outs
+  // itab(pcols), jtab(pcols)
+  // ttab(pcols), utab(pcols)
+  // coef(pcols,ncoef) !coefficient interpolated bilinearly
+  // FIXME; maybe we need to loop over cols
+  constexpr Real one = 1.0;
+  if (itab_1 <= 0.0) {
+    // compute factors for the real part
+    compute_factors(prefr, ref_real, ref_real_tab, itab, ttab);
+
+    // !compute factors for the imaginary part
+    compute_factors(prefi, ref_img, ref_img_tab, jtab, utab);
+  } // itab_1 < -1
+
+  const Real tu = ttab * utab;
+  const Real tuc = ttab - tu;
+  const Real tcuc = one - tuc - utab;
+  const Real tcu = utab - tu;
+  // FIXME: check Fortran to C++ indexing
+  const int jp1 = haero::min(jtab + 1, prefi - 1);
+  const int ip1 = haero::min(itab + 1, prefr - 1);
+  for (int icoef = 0; icoef < ncoef; ++icoef) {
+    coef[icoef] = tcuc * table(icoef, itab, jtab) +
+                  tuc * table(icoef, ip1, jtab) + tu * table(icoef, ip1, jp1) +
+                  tcu * table(icoef, itab, jp1);
+  } // icoef
+
+} // binterp
+
+KOKKOS_INLINE_FUNCTION
+void compute_calcsize_and_water_uptake_dr(
+    const Real &pmid, const Real &temperature, Real &cldn,
+    Real *state_q_kk,   // in
+    const Real *qqcw_k, // in
+    const Real &dt,
+    // outputs
+    int nspec_amode[ntot_amode],
+    int lspectype_amode[ndrop::maxd_aspectype][ntot_amode],
+    Real specdens_amode[ndrop::maxd_aspectype],
+    int lmassptr_amode[ndrop::maxd_aspectype][ntot_amode],
+    Real spechygro[ndrop::maxd_aspectype], Real mean_std_dev_nmodes[ntot_amode],
+    Real dgnumwet_m_kk[ntot_amode], Real qaerwat_m_kk[ntot_amode]) {
+
+  const bool do_adjust = true;
+  const bool do_aitacc_transfer = true;
+  const bool update_mmr = false;
+
+  int numptr_amode[ntot_amode];
+  int mam_idx[ntot_amode][ndrop::nspec_max];
+  int mam_cnst_idx[ntot_amode][ndrop::nspec_max];
+
+  ndrop::get_e3sm_parameters(nspec_amode, lspectype_amode, lmassptr_amode,
+                             numptr_amode, specdens_amode, spechygro, mam_idx,
+                             mam_cnst_idx);
+
+  // FIXME: inv_density: we have different order of species in mam4xx.
```

Comment:
Is this an issue ^^^?

Reply:
I will make an issue for this.

Reply:
See [issue 274](https://github.com/eagles-project/mam4xx/issues/274#issue-2063096769)